### PR TITLE
Towards #2709 - Add DiscoveryInfo to IpAddress

### PR DESCRIPTION
**Pull request by gkleiman**
Status:

- [x] Add DiscoveryInfo Protocol Buffers to marathon.proto
- [x] Add "discoveryInfo" field to IpAddress
- [x] Add DiscoveryInfo JSON serialization
- [x] Add DiscoveryInfo to AppDefinition.json
- [x] Make TaskBuilder fill DiscoveryInfo (PR #2741)
- [x] Add documentation
- [ ] Testing (E2E-Tests)

### CODE REVIEW

File: docs/docs/rest-api/mesosphere/marathon/api/v2/AppDefinition.json [hunk 1]
```diff
@@ -228,6 +228,37 @@
             "type": "object",
```
Comment (aquamatthias) [inline docs/docs/rest-api/mesosphere/marathon/api/v2/AppDefinition.json]: Delete this file.

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 2]
```diff
@@ -239,13 +239,47 @@ trait ContainerFormats {
 trait IpAddressFormats {
   import Formats._
 
-  implicit lazy val NetworkProtocolFormat: Format[mesos.NetworkInfo.Protocol] =
-    enumFormat(mesos.NetworkInfo.Protocol.valueOf, str => s"$str is not a valid protocol")
+  private[this] lazy val ValidPortProtocol: Reads[String] = {
+    implicitly[Reads[String]]
+      .filter(ValidationError("Invalid protocol. Only 'udp' or 'tcp' are allowed."))(
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Mesos proto is not strict here. But I don't see any reason why not to be strict here.

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 3]
```diff
@@ -172,6 +174,24 @@ class TaskBuilder(app: AppDefinition,
...
+        builder.setName(app.id.toHostname)
+        // TODO(gkleiman@) define the default visibility
+        builder.setVisibility(org.apache.mesos.Protos.DiscoveryInfo.Visibility.FRAMEWORK)
+
+        val portsProto = org.apache.mesos.Protos.Ports.newBuilder
+        portsProto.addAllPorts(ports.map(_.toProto).asJava)
+        builder.setPorts(portsProto)
+
+        Some(builder.build)
+      case _ => None
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: If I remember correctly, we decided to not fill DiscoveryInfo in step 1 for non network isolated tasks. Please add a TODO that states that.

**Comment by aquamatthias** [comment]
Looks good! :+1: 
I think (E2E-Tests) is something we can not do very well with the setup we have.
Would really like to have mesosDNS setup to see this in action.

**Comment by aquamatthias** [comment]
Do you want me to merge?

**Comment by gkleiman** [comment]
@aquamatthias thanks a lot for your review. I addressed your comments.

#2744 has to be merged before this one.

**Comment by gkleiman** [comment]
@aquamatthias, @kolloch: branch squashed and ready for final review.

Note that this PR has gk/ip_address as base, to make the review process easier, but it should be merged into gk/ip_per_task after #2744 is merged.

**Comment by aquamatthias** [comment]
+1 from my side.

**Pull request closed without merging**